content/advent-2016/enigma-emulator-in-go.md

```diff
@@ -0,0 +1,486 @@
+
+- "jkm ... glp" (Kenngruppen) are the key groups: they were used to communicate the starting positions of the rotors. [The procedure for using keys](http://users.telenet.be/d.rijmenants/en/enigmaproc.htm) has nothing to do with the Enigma machine itself — the operators did this part — so the only part that matters is being able to set the initial rotor positions.
+
+Now that we know the settings that were used, let's take a look at the constructor:
+
+```
+type RotorConfig struct {
+	ID    string
+	Start byte
+	Ring  int
+}
+
+func NewEnigma(rotorConfiguration []RotorConfig, refID string, plugs []string) *Enigma {
+	// More magic!
+	// 1. Get the rotor models by their UIDs ("I" to "VIII") and load the parameters.
+	// 2. Create a plugboard mapping from the supplied pairs.
+	// 3. Get the reflector by its UID.
+	// 4. Return the new Enigma instance.
+}
+```
+
+I know there is still a lot of magic and boilerplates, but bear with me: including the implementation details would be too confusing at this point. Think of it as a progressive JPEG kind of thing.
+
+A lot of real code in the next section though!
+
+
+
+# Rotors
+
+![](/postimages/advent-2016/enigma-emulator-in-go/rotors.jpg)
+
+Rotors were the heart of the Enigma machines. Essentially, each rotor performs a simple letter subsitution according to its internal wiring:
+
+```
+ABCDEFGHIJKLMNOPQRSTUVWXYZ
+||||||||||||||||||||||||||
+EKMFLGDQVZNTOWYHXUSPAIBRCJ
+```
+
+When the signal goes towards a reflector (from right to left), "A" is encoded to "E", and then the signal goes on to the next rotor. After the reflector, the signal gets reversed: "A" would be encrypted to "U".
+
+As mentioned before, the rotor parts are movable. When the rotor is moved, the substitution table gets offset:
+
+```
+BCDEFGHIJKLMNOPQRSTUVWXYZA
+||||||||||||||||||||||||||
+EKMFLGDQVZNTOWYHXUSPAIBRCJ
+```
+
+Now it is "B", not "A", that encodes to "E".
+
+The operator could set the starting position for each rotor manually, and just like every other configuration parameter, starting positions were changed daily during the war.
+
+A _ring setting_ (set through a small ring on the rotor) would add a fixed offset: with the starting position "B" and the ring setting 4, the table would have an offset of three positions.
+
+Another important part is stepping — moving rules for the rotors. We will get to it after we are done with the rest of the parts, but for now it is important to know that rotors had notches next to one of the letters (rotors V through VIII had two), and when Enigma "encountered" that notch, the rotor on the left was shifted. On the picture above, the rotor on the left has a visible notch at H, and the rotor on the right has a notch at U.
+
+Let's create a simple type for the rotors:
+
+```
+type Rotor struct {
+	ID          string
+	StraightSeq [26]int
+	ReverseSeq  [26]int
+	Turnover    []int
+
+	Offset int
+	Ring   int
+}
+
+type Rotors []*Rotor
```

Comment:
Not certain, so let me try to do it a little differently and check :) I do agree that `[]Rotor` would seem more logical, and I don't entirely remember the full train of thought (and benchmarks) that turned it into the list of pointers.